concept/type_/constraint.rs

```diff
@@ -4,139 +4,544 @@
-    }
-
-    pub fn compute_unique<'a, A: Into<Annotation> + Clone + 'a>(
-        annotations: impl IntoIterator<Item = &'a A>,
-    ) -> Option<AnnotationUnique> {
-        annotations
-            .into_iter()
-            .filter_map(|annotation| match annotation.clone().into() {
-                // Unique and Key cannot be set together
-                Annotation::Unique(_) | Annotation::Key(_) => Some(AnnotationUnique),
-                _ => None,
-            })
-            .next()
-    }
-
-    pub fn compute_cardinality<'a, A: Into<Annotation> + Clone + 'a>(
-        annotations: impl IntoIterator<Item = &'a A>,
-        default: Option<AnnotationCardinality>,
-    ) -> Option<AnnotationCardinality> {
-        let cardinality = annotations
-            .into_iter()
-            .filter_map(|annotation| match annotation.clone().into() {
-                // Cardinality and Key cannot be set together
-                Annotation::Cardinality(card) => Some(card),
-                Annotation::Key(_) => Some(AnnotationKey::CARDINALITY),
-                _ => None,
-            })
-            .next();
-        cardinality.or(default)
-    }
-
-    compute_constraint_one_to_one_annotation!(compute_regex, Annotation::Regex, AnnotationRegex);
-    compute_constraint_one_to_one_annotation!(compute_cascade, Annotation::Cascade, AnnotationCascade);
-    compute_constraint_one_to_one_annotation!(compute_range, Annotation::Range, AnnotationRange);
-    compute_constraint_one_to_one_annotation!(compute_values, Annotation::Values, AnnotationValues);
-
-    pub(crate) fn annotations_by_validation_modes<'a, A: Into<Annotation> + Clone + 'a>(
-        annotations: impl IntoIterator<Item = A>,
-    ) -> Result<HashMap<&'static ConstraintValidationMode, HashSet<Annotation>>, AnnotationError> {
-        let mut map: HashMap<&ConstraintValidationMode, HashSet<Annotation>> = HashMap::new();
-
-        for annotation in annotations.into_iter() {
-            let annotation = annotation.clone().into();
-            let modes = Self::inherited_constraint_validation_mode(annotation.category());
-
-            for mode in modes {
-                map.entry(mode).or_default().insert(annotation.clone());
+macro_rules! unwrap_constraint_description_methods {
+    ($(
+        fn $method_name:ident() -> $return_type:ident = $target_enum:ident;
+    )*) => {
+        $(
+            pub(crate) fn $method_name(self) -> Result<$return_type, ConstraintError> {
+                with_constraint_description!(self, $target_enum, Err(ConstraintError::CannotUnwrapConstraint(stringify!($target_enum).to_string())), |constraint| Ok(constraint.clone()))
             }
+        )*
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum ConstraintCategory {
+    Abstract,
+    Distinct,
+    Independent,
+    Unique,
+    Cardinality,
+    Regex,
+    Range,
+    Values,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub enum ConstraintDescription {
```

Comment:
Generally, I like what I've got with constraints, but passing `ConstraintDescription` around and just calling "vallidate it, please" is not statically checked by Rust (in a function to check Regex constraints, I could potentially call "get all Card constraints, validate" and it would be fine. Same for general getters of Constraints. I'm sure it could be refactored to consider static checks, but I didn't have that much motivation to do so because we're covered by BDD tests from all sides like crazy, and they should do the job. Still, something important to highlight